base_layer/core/src/base_node/comms_interface/inbound_handlers.rs

```diff
@@ -267,6 +269,23 @@ where B: BlockchainBackend + 'static
             NodeCommsRequest::GetNewBlockTemplate(request) => {
                 let best_block_header = self.blockchain_db.fetch_tip_header().await?;
                 let mut last_seen_hash = self.mempool.get_last_seen_hash().await?;
+                {
+                    let read_lock = self.cached_block_template.read().await;
+                    if let Some(cache) = read_lock.as_ref() {
+                        if cache.0 == last_seen_hash && &cache.1.header.prev_hash == best_block_header.hash() {
+                            return Ok(NodeCommsResponse::NewBlockTemplate(cache.1.clone()));
+                        }
+                    }
+                }
```

Comment:
I would remove these lines - we are not supposed to access this function asynchronously. 
- If one thead has obtained the write lock:
   - the next thread will wait at the read lock.
   - the mempool has changed and the next thread should wait at the read lock.
The lines below do exactly the same thing.